Implemented support of storing cell calculated values

Pankraty opened the pull request:
This PR implements requirements from #667. 

The idea is the following:

1. ```XLWorkbook``` has a property RecalculationCounter that is increased on every modification (either cell value or cell formula is modified).
2. ```XLCell``` has a few new properties:
- ```ValueCalculated``` - stores the evaluated value of cell formula.
- ```EvaluatedAtVersion``` - stores the value of workbook's ```RecalculationCounte```r that it had at the moment of cell last evaluation. If this value equals the current value of ```RecalculationCounter``` it indicates that ```ValueCalculated``` stores correct value and no re-evaluation has to be performed.
- ```EditedAtVersion``` - stores the value of workbook's ```RecalculationCounter``` that it had at the moment of cell last modification (value or formula). If this value is greater than ```EvaluatedAtVersion``` then cell needs re-evaluation, as well as all dependent cells do.
- ```RecalculationNeeded``` - getter of this property checks if the cell needs to be re-evaluated. It can be true if the cell was modified from the moment of previous evaluation (```EditedAtVersion > EvaluatedAtVersion```) or if any of the cell the current one depends on requires re-evaluation.
3. ```XLCalcEngine``` is extended with methods for enumerating ranges and cells that affect the specified expression directly. To get the desired ranges it simply calls a ```Parse``` method and collects all CellRangeReferences that are included in the expression. To avoid duplication two custom equality comparers (XLRangeAddressComparer and XLAddressComparer) has been introduced.

With these changes, the calculation becomes much more efficient. And the modification of a single cell causes the deferred re-evaluation of dependent cells, leaving the rest untouched.

I covered many scenarios with tests, and ready to give additional comments if needed.

#### Questions:
- Is there a blog post? — No
- Does the knowledge base need an update? — Yes, I suppose so
- Does this add new (C#) dependencies? — No

- [ ] C# Code Review: @csreviewer
- [ ] Test Automation Review: @csreviewer

igitur commented:
Please rebase on latest `develop`.

igitur commented:
What's the pros and cons of changing the `RecalculationCounter` from type `long` to type `DateTime`. I know we've briefly chatted about this before. I think I prefer `DateTime`. It would work the same, except that it has the added benefit of actually showing you when the calculations where last redone. A normal `long` counter can't do that. What do you think

vbjay commented:
Why not use long but make it store ticks?  Convert to datetime if desired?

On Fri, Mar 23, 2018 at 1:01 PM Francois Botha <notifications@github.com>
wrote:

> What's the pros and cons of changing the RecalculationCounter from type
> long to type DateTime. I know we've briefly chatted about this before. I
> think I prefer DateTime. It would work the same, except that it has the
> added benefit of actually showing you when the calculations where last
> redone. A normal long counter can't do that. What do you think
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/679#issuecomment-375733484>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsWHWKDfzRua2C0QX3nTBMctnKYbmks5thSpkgaJpZM4R4swa>
> .
>

vbjay commented:
Heck, we could have a readonly property that is the datetime based off the
RecalculationCounter value.  Could even be nullable to say a recalc wasn't
done yet.

On Fri, Mar 23, 2018 at 2:50 PM Jay Asbury <vbjay.net@gmail.com> wrote:

> Why not use long but make it store ticks?  Convert to datetime if desired?
>
> On Fri, Mar 23, 2018 at 1:01 PM Francois Botha <notifications@github.com>
> wrote:
>
>> What's the pros and cons of changing the RecalculationCounter from type
>> long to type DateTime. I know we've briefly chatted about this before. I
>> think I prefer DateTime. It would work the same, except that it has the
>> added benefit of actually showing you when the calculations where last
>> redone. A normal long counter can't do that. What do you think
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/ClosedXML/ClosedXML/pull/679#issuecomment-375733484>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ADdhsWHWKDfzRua2C0QX3nTBMctnKYbmks5thSpkgaJpZM4R4swa>
>> .
>>
>

Pankraty commented:
I don't mind adding ```DateTime``` property on a workbook or a worksheet level, along with the numeric counter. But using it for determining if the cell needs re-evaluation looks like a time bomb for me, it would strike eventually. 
I have not much to add to my previous reasoning (https://github.com/ClosedXML/ClosedXML/issues/667#issuecomment-360144531). To put it shortly, getting current time value is slower, harder to deal with in multithreading computations; also time sometimes goes backward making a reproduction of some bugs extremely hard.

vbjay commented:
Good points.  A counter solves those problems and if we use interlocked, we
can prevent thread issues.

On Fri, Mar 23, 2018 at 2:57 PM Aleksei <notifications@github.com> wrote:

> I don't mind adding DateTime property on a workbook or a worksheet level,
> along with the numeric counter. But using it for determining if the cell
> needs re-evaluation looks like a time bomb for me, it would strike
> eventually.
> I have not much to add to my previous reasoning (#667 (comment)
> <https://github.com/ClosedXML/ClosedXML/issues/667#issuecomment-360144531>).
> To put it shortly, getting current time value is slower, harder to deal
> with in multithreading computations; also time sometimes goes backward
> making a reproduction of some bugs extremely hard.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/679#issuecomment-375767174>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsarDMJgwg8J6SqtzKNWcN0Dia_nVks5thUWkgaJpZM4R4swa>
> .
>

Pankraty commented:
Rebased

Pankraty commented:
For not spreading a discussion among too many PRs I'll proceed here.
I implemented an alternative approach using ```DateTime``` instead of ```long``` counter and I must admit that some of my concerns were groundless. It does not seem to be any slower and using ```DateTime.UtcNow``` protects us from arbitrary timezone changes and switches to daylight saving time.
The other important issue still exists though. I have experimented a bit with the following code:
```
using (var wb = new XLWorkbook())
{
    var ws = wb.AddWorksheet("Sheet1");

    ws.Cell(1, 1).Value = 1;
    ws.Cell(1, 2).FormulaA1 = "A1*10";
    Console.WriteLine("{2} B1 value = {0}, expected {1}", ws.Cell(1, 2).Value, 10, DateTime.UtcNow);
    Console.WriteLine("Pause...");
    Console.ReadKey();

    ws.Cell(1, 1).Value = 2;
    Console.WriteLine("{2} B1 value = {0}, expected {1}", ws.Cell(1, 2).Value, 20, DateTime.UtcNow);
}
```
During the pause I changed system time settings or let the system to syncronize time with NTP server. The result is quite predictable:
```
4/18/2018 5:28:10 AM B1 value = 10, expected 10
Pause...
4/18/2018 5:26:19 AM B1 value = 10, expected 20
```

What is interesting is that 23 tests running fine on a local machine fail at AppVeyor. Do I miss something essential or time flows strangely at the build server?

vbjay commented:
Read the remarks.
https://msdn.microsoft.com/en-us/library/system.datetime.utcnow(v=vs.110).aspx

On Wed, Apr 18, 2018, 1:54 AM Aleksei <notifications@github.com> wrote:

> For not spreading a discussion among too many PRs I'll proceed here.
> I implemented an alternative approach using DateTime instead of long
> counter and I must admit that some of my concerns were groundless. It does
> not seem to be any slower and using DateTime.UtcNow protects us from
> arbitrary timezone changes and switches to daylight saving time.
> The other important issue still exists though. I have experimented a bit
> with the following code:
>
> using (var wb = new XLWorkbook())
> {
>     var ws = wb.AddWorksheet("Sheet1");
>
>     ws.Cell(1, 1).Value = 1;
>     ws.Cell(1, 2).FormulaA1 = "A1*10";
>     Console.WriteLine("{2} B1 value = {0}, expected {1}", ws.Cell(1, 2).Value, 10, DateTime.UtcNow);
>     Console.WriteLine("Pause...");
>     Console.ReadKey();
>
>     ws.Cell(1, 1).Value = 2;
>     Console.WriteLine("{2} B1 value = {0}, expected {1}", ws.Cell(1, 2).Value, 20, DateTime.UtcNow);
> }
>
> During the pause I changed system time settings or let the system to
> syncronize time with NTP server. The result is quite predictable:
>
> 4/18/2018 5:28:10 AM B1 value = 10, expected 10
> Pause...
> 4/18/2018 5:26:19 AM B1 value = 10, expected 20
>
> What is interesting is that 23 tests running fine on a local machine fail
> at AppVeyor. Do I miss something essential or time flows strangely at the
> build server?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ClosedXML/ClosedXML/pull/679#issuecomment-382270565>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADdhsSLZ8OAmMgmXCtfdjPy-sqLCA-gMks5tptT9gaJpZM4R4swa>
> .
>

igitur commented on ClosedXML/Excel/Cells/IXLCell.cs:
```diff
@@ -310,8 +310,31 @@ public interface IXLCell
 
         IXLCell CopyTo(String target);
 
+        /// <summary>
```
Ok, so for this interface I have some other ideas. 

Firstly, for a user, I think it is going to be confusing to have both `ValueCached` and `ValueCalculated`. I think we should have only 1 and I think ideally it should return `Object`. That will be a breaking change, so maybe we should phase it in with keeping both for now, and marking one as obsolete. We'll need a property name that correctly describes both. Maybe `Object CachedValue {get;}` is OK.

igitur commented on ClosedXML/Excel/Cells/IXLCell.cs:
```diff
@@ -310,8 +310,31 @@ public interface IXLCell
+        /// <summary>
+        /// Flag indicating that previously calculated cell value may be not valid anymore and has to be re-evaluated.
+        /// </summary>
+        Boolean RecalculationNeeded { get; }
```
Please change to `NeedsRecalculation`. I prefer `VerbObject` format (active voice) for property names, instead of passive voice.

igitur commented on ClosedXML/Excel/Cells/IXLCell.cs:
```diff
@@ -310,8 +310,31 @@ public interface IXLCell
+        /// </summary>
+        /// <param name="force">Flag indicating whether a recalculation must be performed even is cell does not need it.</param>
+        /// <returns>Null if cell does not contain a formula. Calculated value otherwise.</returns>
+        object Evaluate(bool force = false);
```
I don't think this property is necessary. I would drop this and rather have a `void InvalidateFormula()` method. Then, only when the cell's value is requested, is the value calculated again. `Evaluate` can stay internal. Agree?

igitur commented on ClosedXML/Excel/XLWorkbook.cs:
```diff
@@ -119,7 +119,22 @@ public static XLWorkbook OpenFromTemplate(String path)
+        /// Notify that workbook data has been changed which means that cached formula values
+        /// need to be re-evaluated.
+        /// </summary>
+        internal void NotifyRecalculationNeeded()
```
If I understand this correctly, I think we should rename to `InvalidateFormulas`.

igitur commented on ClosedXML/Excel/Cells/XLCell.cs:
```diff
@@ -1199,6 +1251,66 @@ public IXLCell AddToNamed(string rangeName, XLScope scope, string comment)
+        /// The value of <see cref="XLWorkbook.RecalculationCounter"/> that workbook had at the moment of cell last modification.
+        /// If this value is greater than <see cref="EvaluatedAtVersion"/> then cell needs re-evaluation, as well as all dependent cells do.
+        /// </summary>
+        internal long EditedAtVersion { get; private set; }
```
Rather `ModifiedAtVersion` ?

igitur commented on ClosedXML/Excel/Cells/XLCell.cs:
```diff
@@ -362,92 +361,144 @@ public string GetFormattedString()
+        /// <returns>Null if cell does not contain a formula. Calculated value otherwise.</returns>
+        public object Evaluate(bool force = false)
+        {
+            if (NeedsRecalculation || force)
```
`if (force || NeedsRecalculation)` be better to avoid `NeedsRecalculation` to be evaluated when `force == true`.

igitur commented on ClosedXML/Excel/CalcEngine/XLCalcEngine.cs:
```diff
@@ -23,6 +23,50 @@ public XLCalcEngine(IXLWorksheet ws) : this(ws.Workbook)
+        public IEnumerable<IXLCell> GetAffectingCells(string expression)
+        {
+            if (String.IsNullOrWhiteSpace(expression) && String.IsNullOrEmpty(expression))
+                yield break;
```
`yield break` isn't required.

igitur commented on ClosedXML/Excel/CalcEngine/XLCalcEngine.cs:
```diff
@@ -23,6 +23,50 @@ public XLCalcEngine(IXLWorksheet ws) : this(ws.Workbook)
+            }
+        }
+
+        public IEnumerable<IXLCell> GetAffectingCells(string expression)
```
Please rename to `GetPrecedentCells`.

By the way, we might want to expose this method publicly on the API. Could be useful. Leave for later.

igitur commented on ClosedXML/Excel/CalcEngine/XLCalcEngine.cs:
```diff
@@ -49,66 +93,21 @@ public override object GetExternalObject(string identifier)
-                _evaluating = false;
-            }
+            var res = new CellRangeReference(range, this);
+            if (_cellRanges != null)
```
`_cellRanges?.Add(res.Range);`

igitur commented on ClosedXML/Excel/CalcEngine/XLCalcEngine.cs:
```diff
@@ -23,6 +23,50 @@ public XLCalcEngine(IXLWorksheet ws) : this(ws.Workbook)
+        /// </summary>
+        /// <param name="expression">Formula to parse.</param>
+        /// <returns>Collection of ranges included into the expression.</returns>
+        public IEnumerable<IXLRange> GetAffectingRanges(string expression)
```
Rename to `GetPrecedentRanges`. I'm trying to align with Excel:

![image](https://user-images.githubusercontent.com/145854/38986895-2b895e96-43ce-11e8-9db1-1bb6fe5d3dd2.png)

Pankraty commented on ClosedXML/Excel/CalcEngine/XLCalcEngine.cs:
```diff
@@ -23,6 +23,50 @@ public XLCalcEngine(IXLWorksheet ws) : this(ws.Workbook)
+
+        public IEnumerable<IXLCell> GetAffectingCells(string expression)
+        {
+            if (String.IsNullOrWhiteSpace(expression) && String.IsNullOrEmpty(expression))
```
This condition is really weird, isn't it?

igitur replied:
Sorry, I don't understand. What I meant was that you could just do this:
```c#
if (!String.IsNullOrWhiteSpace(expression) || !String.IsNullOrEmpty(expression))
{
    var ranges = GetAffectingRanges(expression);
    var visitedCells = new HashSet<IXLAddress>(new XLAddressComparer(true));
    var cells = ranges.SelectMany(range => range.Cells()).Distinct();
    foreach (var cell in cells)
    {
        if (!visitedCells.Contains(cell.Address))
        {
            visitedCells.Add(cell.Address);
            yield return cell;
        }
    }
}
```

igitur replied:
Oh, do you mean we could just do `if (!String.IsNullOrWhiteSpace(expression))` ?

Pankraty replied:
Exactly.

igitur replied:
Ok, so then that method's body could be just:
```c#
if (!String.IsNullOrWhiteSpace(expression))
{
    var ranges = GetAffectingRanges(expression);
    var visitedCells = new HashSet<IXLAddress>(new XLAddressComparer(true));
    var cells = ranges.SelectMany(range => range.Cells()).Distinct();
    foreach (var cell in cells)
    {
        if (!visitedCells.Contains(cell.Address))
        {
            visitedCells.Add(cell.Address);
            yield return cell;
        }
    }
}
```

igitur commented:
Great work, again! Thanks.
